keps/sig-node/5067-pod-generation/README.md

```diff
@@ -0,0 +1,1006 @@
+sole writer of `status.observedGeneration`. 
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Today, it is possible for a client to custom-set a pod's `metadata.generation`, but once
+set, the `metadata.generation` cannot be updated. That means that there is a theoretical 
+possibility for an external client to be setting `metadata.generation` on pod 
+create and depend on that fixed value somehow in its own reconciliation logic.
+
+That said, the `metadata.generation` field is described as ["set by the system and monotonically increasing"](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata),
+and thus it should be clear enough that `metadata.generation` was not intended to be
+used in this way.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+#### API server
+
+For a newly created pod, the API server will set `metadata.generation` to 1. For any updates
+to the [PodSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#podspec-v1-core), the API server will increment `metadata.generation` by 1. The logic to set new pods'
+`metadata.generation` to 1 and to increment `metadata.generation` on update will
+run after all mutating webhooks have finished. 
+
+The pod `metadata.generation` will also continue to be incremented on graceful delete or 
+deferred delete, just as the API server currently does for pods and other 
+objects today.
+
+Pod updates that would not result in `metadata.generation` being incremented include:
+* Changes to metadata. This means that if a Pod uses the downward API to make
+pod metadata available to containers, Pod behavior can change without the
+generation being incremented. We will consider this working as intended.
+* Changes to status.
+
+Any attempts by clients to set or modify the `metadata.generation` field themselves will be
+ignored and overridden by the API Server. This is consistent with existing 
+behavior of the `metadata.generation` field in all other objects. 
+
+The API server will have special-case code to avoid setting or updating
+`metadata.generation` for mirror pods. 
+
+#### Kubelet
+
+When the Kubelet updates the pod status as part of the [pod sync loop](https://github.com/kubernetes/kubernetes/blob/45d0fddaf1f24f7b559eb936308ce2aeb9871850/pkg/kubelet/pod_workers.go#L1214), 
```

Comment:
Hmm........ this is more about implementation detail, and my recollection of the logic might be completely wrong, but brining it up here just in case. 

I think pod workers check the current running containers and update the status at the *beginning* of a sync, before taking any actions. So the pod spec it possesses at that time hasn't been actuated yet. The status reflects the previous sync in the regular cases (except for a few cases where status updates are triggered first). 

I could be completely wrong but please double check this.  @natasha41575 @tallclair

Reply:
We could cache the generation version that was actually used to actuate the pod, but this won't survive kubelet restarts. So we might be looking at corner cases (restarts) until we can revamp the kubelet internals.

Reply:
Discussed with @tallclair on this some more. Leaving a comment here for the record. @tallclair will likely follow up with a better written explanation :) 

We can simply use the pod spec generation observed at the time of the creating/reporting the pod status, even though kubelet hasn't actuated the change yet.

 - What we would be able to capture is kubelet's explicit decision to admit/reject a change, and/or the acknowledge of a new addition in the spec (e.g., ephemeral container added, but in the waiting state).
 - What we won't be able to reflect directly is whether kubelet has acted on the update of certain mutatble fields (e.g., container image). But that's okay for now.

Reply:
Here's a field-by-field analysis: https://docs.google.com/document/d/1-4CR2NmDJotCM13YQyr8OEeGIczlc3WFmV1BhdUrWcw/edit?tab=t.0

The conclusion is sort of that current generation makes sense for some fields, and previous generation is more relevant for other fields.

In my opinion, it still makes sense to report the latest generation the Kubelet has seen, and just be careful to document the nuance that it is not necessarily the generation reflected by every field in the status.

Reply:
I think the root of the problem is that the Kubelet is observing 2 different sources, and reporting on both of them in the pod status: the k8s API, and the container runtime. The status associated with observations on the k8s API will always be the current generation, and the observations on the runtime will be the generation from the previous pod sync (or potentially older). I'm not sure it's worth adding a separate field to track the latter.

Reply:
> I'm not sure it's worth adding a separate field to track the latter.

Agree it's not worth adding another field. But maybe if/when we change kubelet's internal syncing logic, the meaning/nuances of this field will change. As long as we make sure that the field only indicates that the kubelet has seen the spec, and doesn't necessarily mean it has tried to act on it, it should be okay :)

Reply:
> What we won't be able to reflect directly is whether kubelet has acted on the update of certain mutatble fields (e.g., container image). But that's okay for now.
> ...
> meaning/nuances of this field will change.

We need to document the field very generically and explicitly state that it cannot be used to assume what the value of observerGeneration will be comparing to other fields. What would be the promise for this field so it will be useful for the in-place resize KEP? Can we list it very explicitly in the docs? Especially if we plan to change it in future. The field like this is so tempting to start using it for a consistency check.

Reply:
> What would be the promise for this field so it will be useful for the in-place resize KEP? 

@tallclair can correct me if I'm wrong, but IIUC the purpose of this field (with regards to the in-place resize KEP) is to answer the question "has the kubelet seen this podspec generation yet"? So I agree that it makes sense to report the latest generation that kubelet has seen, even if that is not necessarily a reflection of every field in the status.

> Can we list it very explicitly in the docs?

Yes, I can add a note that we need to document this very clearly and carefully.

Reply:
I added the discussion here to the body of the KEP - a description of what `observedGeneration` can capture vs what it can't, and that we need to document it carefully